pageserver/src/tenant/timeline.rs

```diff
@@ -1221,70 +1232,161 @@ impl Timeline {
-            task_mgr::spawn(
-                task_mgr::BACKGROUND_RUNTIME.handle(),
-                task_mgr::TaskKind::InitialLogicalSizeCalculation,
-                Some(self.tenant_id),
-                Some(self.timeline_id),
-                "initial size calculation",
-                false,
-                async move {
-                    let mut timeline_state_updates = self_clone.subscribe_for_state_updates();
-                    let self_calculation = Arc::clone(&self_clone);
-                    tokio::select! {
-                        calculation_result = spawn_blocking(move || self_calculation.calculate_logical_size(init_lsn)) => {
-                            let calculated_size = calculation_result
-                                .context("Failed to spawn calculation result task")?
-                                .context("Failed to calculate logical size")?;
-                            match self_clone.current_logical_size.initial_logical_size.set(calculated_size) {
-                                Ok(()) => info!("Successfully calculated initial logical size"),
-                                Err(existing_size) => error!("Tried to update initial timeline size value to {calculated_size}, but the size was already set to {existing_size}, not changing"),
-                            }
-                            Ok(())
-                        },
-                        new_event = async {
-                            loop {
-                                match timeline_state_updates.changed().await {
-                                    Ok(()) => {
-                                        let new_state = *timeline_state_updates.borrow();
-                                        match new_state {
-                                            // we're running this job for active timelines only
-                                            TimelineState::Active => continue,
-                                            TimelineState::Broken | TimelineState::Stopping | TimelineState::Suspended => return Some(new_state),
-                                        }
-                                    }
-                                    Err(_sender_dropped_error) => return None,
-                                }
-                            }
-                        } => {
-                            match new_event {
-                                Some(new_state) => info!("Timeline became inactive (new state: {new_state:?}), dropping current connections until it reactivates"),
-                                None => info!("Timeline dropped state updates sender, stopping init size calculation"),
+        use InitialLogicalSizeComputationState::*;
+        let mut guard = self.initial_size_computation_state.lock().unwrap();
+        match *guard {
+            Running | Success => return,
+            NotStarted | FailedWillRetryNextTime => *guard = Running,
+        }
+        drop(guard);
```

Comment:
I wonder if we could get simplify this a bit (mutex, state enum) with a single `Arc<tokio::sync::Semaphore>` with one permit. Then do a check for `Semaphore::try_acquire_owned()` together with `self.current_logical_size.initial_logical_size.get().is_none()`, and only then start the calculation?

Reply:
The behavior we want to
1. at most one concurrent initial logical size computation for the timeline at a time
2. after the first initial logical size computation succeeds, don't start new ones

So, with the semaphore, we'd need to use `OwnedSemaphorePermit::forget()`.

IDK if this is better in terms of readability than a state enum.

I was contemplating of using `atomic_enum` + compare and swap (like with the `AtomicBool` that we had before this change).
That 

I'll spend a few minutes drafting out your proposal.

Reply:
Pushed commit

```
fixup [4/4]: use Semaphore + forgetting permit instead of explicit state enum
```

I'm OK with it, don't think it's particularly readable though.

Reply:
I think it's shorter, so I enjoy it :)